src/index.ts

```diff
@@ -284,377 +201,289 @@ export interface TOTPStrategyOptions {
-          }
-          return await this.failure(
-            this.customErrors.inactiveTotp || ERRORS.INACTIVE_TOTP,
-            request,
-            sessionStorage,
-            options,
-            error,
-          )
-        }
-
-        return await this.failure(error.message, request, sessionStorage, options, error)
+      const code = formDataTotp ?? this._getMagicLinkCode(request)
+      if (code) {
+        if (!sessionEmail || !sessionTotp) throw new Error(this.customErrors.expiredTotp)
+        await this._validateTOTP({
+          code,
+          sessionTotp: sessionTotp,
+          session,
+          sessionStorage,
+          options,
+        })
+
+        // Allow developer to handle user validation.
+        const user = await this.verify({
+          email: sessionEmail,
+        })
+
+        session.set(options.sessionKey, user)
+        session.unset(this.sessionEmailKey)
+        session.unset(this.sessionTotpKey)
+        session.unset(options.sessionErrorKey)
+
+        throw redirect(options.successRedirect, {
+          headers: {
+            'set-cookie': await sessionStorage.commitSession(session, {
+              maxAge: this.maxAge,
+            }),
+          },
+        })
       }
-      if (typeof error === 'string') {
+      throw new Error(this.customErrors.requiredEmail)
+    } catch (throwable) {
+      if (throwable instanceof Response) throw throwable
+      if (throwable instanceof Error) {
         return await this.failure(
-          error,
+          throwable.message,
           request,
           sessionStorage,
           options,
-          new Error(error),
+          throwable,
         )
       }
-      return await this.failure(
-        ERRORS.UNKNOWN_ERROR,
-        request,
-        sessionStorage,
-        options,
-        new Error(JSON.stringify(error, null, 2)),
-      )
+      throw throwable
     }
-
-    if (!user) throw new Error(ERRORS.USER_NOT_FOUND)
-
-    return this.success(user, request, sessionStorage, options)
-  }
-
-  private async _validateEmailDefaults(email: string) {
-    const regexEmail = /^[^\s@]+@[^\s@]+\.[^\s@]+$/gm
-    if (!regexEmail.test(email)) throw new Error(this.customErrors.invalidEmail)
   }
 
-  private async _validateTOTP(sessionTotp: string, otp: string, expiresAt: Date) {
-    // Retrieve encrypted TOTP from database.
-    const dbTOTP = await this.readTOTP(sessionTotp)
-    if (!dbTOTP || !dbTOTP.hash) throw new Error(this.customErrors.totpNotFound)
-
-    if (dbTOTP.active !== true) {
-      throw new Error(this.customErrors.inactiveTotp)
+  private async _generateAndSendTOTP({
+    email,
+    session,
+    sessionStorage,
+    request,
+    formData,
+    options,
+  }: {
+    email: string
+    session: Session
+    sessionStorage: SessionStorage
+    request: Request
+    formData: FormData
+    options: RequiredAuthenticateOptions
+  }) {
+    if (!(await this.validateEmail(email))) {
```

Comment:
Although awaiting in `if` statements is quite common, I rarely use them and also find it hard to determine the outcome, whether it's void or a boolean.

Could we also avoid this and refactor it by initializing the value with a constant and checking for it afterward?